packages/cli/src/services/FileCommandLoader.ts

```diff
@@ -0,0 +1,215 @@
+ * - Adapting valid definitions into executable SlashCommand objects.
+ * - Handling file system errors and malformed files gracefully.
+ */
+export class FileCommandLoader implements ICommandLoader {
+  private readonly projectRoot: string;
+
+  constructor(private readonly config: Config | null) {
+    this.projectRoot = config?.getProjectRoot() || process.cwd();
+  }
+
+  /**
+   * Loads all commands, applying the precedence rule where project-level
+   * commands override user-level commands with the same name.
+   * @param signal An AbortSignal to cancel the loading process.
+   * @returns A promise that resolves to an array of loaded SlashCommands.
+   */
+  async loadCommands(signal?: AbortSignal): Promise<SlashCommand[]> {
+    const userDir = getUserCommandsDir();
+    const projectDir = getProjectCommandsDir(this.projectRoot);
+
+    const [userCommands, projectCommands] = await Promise.all([
+      this.loadCommandsFromDir(userDir, userDir, signal),
+      this.loadCommandsFromDir(projectDir, projectDir, signal),
+    ]);
+
+    const commandMap = new Map<string, SlashCommand>();
+    for (const cmd of userCommands) {
+      commandMap.set(cmd.name, cmd);
+    }
+    for (const cmd of projectCommands) {
+      commandMap.set(cmd.name, cmd);
+    }
+
+    return Array.from(commandMap.values());
+  }
+
+  /**
+   * Recursively scans a directory for .toml files and adapts them into commands.
+   * @param baseDir The root command directory (e.g., ~/.gemini/commands) used
+   *   to calculate the command's relative path name.
+   * @param currentDir The directory currently being scanned.
+   * @param signal An AbortSignal to cancel the loading process.
+   * @returns A promise resolving to an array of commands found in the directory.
+   */
+  private async loadCommandsFromDir(
+    baseDir: string,
+    currentDir: string,
+    signal?: AbortSignal,
+  ): Promise<SlashCommand[]> {
+    if (signal?.aborted) {
+      return [];
+    }
+
+    try {
+      const entries = await fs.readdir(currentDir, { withFileTypes: true });
+      const commands: SlashCommand[] = [];
+
+      for (const entry of entries) {
+        if (signal?.aborted) break;
+
+        const fullPath = path.join(currentDir, entry.name);
+        if (entry.isDirectory()) {
+          const subCommands = await this.loadCommandsFromDir(
+            baseDir,
+            fullPath,
+            signal,
+          );
+          commands.push(...subCommands);
+        } else if (entry.isFile() && entry.name.endsWith('.toml')) {
+          const command = await this.parseAndAdaptFile(fullPath, baseDir);
+          if (command) {
+            commands.push(command);
+          }
+        }
+      }
+      return commands;
+    } catch (error: unknown) {
+      if (isErrnoException(error) && error.code !== 'ENOENT') {
+        console.error(
+          `[FileCommandLoader] Error reading directory ${currentDir}:`,
+          error.message,
+        );
+      }
```

Comment:
With this it looks like we silence any non Errno exceptions, what's the motivation there? Ditto q elsewhere